File: 2023/js/day09.js
data = `0 3 6 9 12 15
1 3 6 10 15 21
10 13 16 21 30 45`;

interpolate_prev = (seq) => {
  let iters = [];
  let iter = seq.slice(1).map((x, i) => x - seq[i]);
  while (!iter.every((x) => x === 0)) {
    iters.push(iter);
    iter = iter.slice(1).map((x, i) => x - iter[i]);
  }
  let diff = iters.reverse().reduce((a, ii) => ii[0] - a, 0);
    console.log(seq, seq[0] - diff);
  return seq[0] - diff;
};

interpolate_next = (seq) => {
  let iters = [];
  let iter = seq.slice(1).map((x, i) => x - seq[i]);
  while (!iter.every((x) => x === 0)) {
    iters.push(iter);
    iter = iter.slice(1).map((x, i) => x - iter[i]);
  }
  let diff = iters.reverse().reduce((a, ii) => ii[ii.length - 1] + a, 0);
  return diff + seq[seq.length - 1];
};

ans = data.split`\n`
  .map((x) => x.split` `.map((y) => +y))
  .map(interpolate_prev)
  .reduce((a, b) => a + b, 0);
console.log(ans);

File: 2023/rust/src/day09.rs
use itertools::Itertools;

pub fn sln(input: String, q: Option<u8>) {
    let parsed: Vec<Vec<i64>> = input
        .lines()
        .map(|l| {
            l.split_whitespace()
                .map(|n| n.parse::<i64>().unwrap())
                .collect::<Vec<i64>>()
        })
        .collect();
    match q {
        Some(1) => println!("{}", q1(&parsed)),
        Some(2) => println!("{}", q2(&parsed)),
        _ => {
            println!("{}", q1(&parsed));
            println!("{}", q2(&parsed));
        }
    }
}

enum Dir {
    Prev,
    Next,
}

fn interpolate(seq: &Vec<i64>, dir: Dir) -> i64 {
    let mut iters: Vec<Vec<i64>> = Vec::new();
    let mut flag = false;
    let mut iter = seq.clone();
    while !flag {
        iter = iter
            .iter()
            .tuple_windows()
            .map(|(curr, next)| next - curr)
            .collect();
        flag = iter.iter().all_equal();
        iters.push(iter.clone());
    }
    let fun = |diff: i64, list: &Vec<i64>| -> i64 {
        match dir {
            Dir::Prev => list[0] - diff,
            Dir::Next => list.iter().last().unwrap() + diff,
        }
    };

    let diff = iters.iter().rev().fold(0i64, fun);
    fun(diff, seq)
}

fn q1(input: &[Vec<i64>]) -> i64 {
    input
        .iter()
        .fold(0, |a, seq| a + interpolate(seq, Dir::Next))
}

fn q2(input: &[Vec<i64>]) -> i64 {
    input
        .iter()
        .fold(0, |a, seq| a + interpolate(seq, Dir::Prev))
}
